lib/SILGen/SILGenExpr.cpp

```diff
@@ -3399,7 +3399,11 @@ static ManagedValue emitKeyPathRValueBase(SILGenFunction &subSGF,
   
   // Upcast a class instance to the property's declared type if necessary.
   if (auto propertyClass = storage->getDeclContext()->getSelfClassDecl()) {
-    if (baseType->getClassOrBoundGenericClass() != propertyClass) {
+    auto baseClass = baseType->getClassOrBoundGenericClass();
+    if (auto dynamicSelfType = baseType->getAs<DynamicSelfType>()) {
+      baseClass = dynamicSelfType->getSelfType()->getClassOrBoundGenericClass();
```

Comment:
Instead of calling getClassOrBoundGenericClass() twice, how about we unwrap the DynamicSelfType first:

```
if (auto selfType = baseType->getAs<DynamicSelfType>()) {
  baseType = selfType->getSelfType();
baseClass = baseType->getClassOrBoundGenericClass();
ASSERT(baseClass != nullptr);
```

The assert means that if this kind of bug comes up again, we get a better crash report than we otherwise would from a segfault.

Reply:
@slavapestov 

Thank you for checking. I tried the below code, but the assertion failed in [`subclass_generics`](https://github.com/swiftlang/swift/blob/f31c89a73e3f14e8f5e7ad1b92dd434ef12654b1/test/SILGen/keypaths.swift#L510).

```
if (auto selfType = baseType->getAs<DynamicSelfType>())
      baseType = selfType->getSelfType()->getCanonicalType(); // Need getCanonicalType() since baseType is CanType 
auto baseClass = baseType->getClassOrBoundGenericClass();
ASSERT(baseClass != nullptr);
```

Am I misunderstanding something?

Reply:
@slavapestov 
As far as I confirmed, `baseClass` still could be nil when `baseType` is `ArchetypeType`.
Please correct me if I'm wrong. Thank you.